src/mixer/doc/emu-config.1
.\" emu-config
.de Id
.ds Dt \\$4
..
.Id $Id: emu-config.1,v 1.3 2002/01/19 06:07:58 dbertrand Exp $
.ds = \-\^\-
.de Sp
.if t .sp .3
.if n .sp
..
.TH emu-config 1 "25 June 2001" "emu10k1-tools" "Emu10k1 Users Manual"
.SH NAME
emu-config \- configuration tool for emu10k1 and Audigy based soundcards.
.SH SYNOPSIS
\fBemu-config\fP \fB-?Vadp\fP -s\fIBUS\fP -r\fIBUS\fP -s\fISOURCE\fP
.SH DESCRIPTION
.B emu-config
is used to configure emu10k1 based soundcards. It can set shared
analog/digital jacks (i.e. the yellow jack) on the '5.1' series of
soundblaster Lives. And it can set the recording modes (more details
below). 

.SH OPTIONS
.TP 3
\fB\-a\fP or \fB\*=analog\fP
Set output in analog mode (SBLive 5.1 and Audigy cards only).
.TP
\fB\-b\fIBUS\fP or \fB\*=bus=\fIBUS\fP        
sets/prints the FX recording bus/es
.TP
\fB\-B\fI<on/off>\fP or \fB\*=boost=\fI<on/off>\fP   
Set a 12dB boost on the analog front out either 'on' or 'off'. This
option is only valid for SBLive cards.  
.TP
\fB\-d\fP or \fB\*=digital\fP
Set output in digital mode (SBLive 5.1 and Audigy cards only).
.TP
\fB\-f\fIFREQ\fP or \fB\*=spdifrate=\fIFREQ\fP
Set the sample rate of the output spdif on Audigy cards
.TP
\fB\-i\fP or \fB\*=ir\fP
Enables Live Drive's and Audigy Drive's ir remote control and midi port.
.TP
\fB\-I\fP or \fB\*=info\fP
Prints info on the card and driver.
.TP
\fB\-m\fIFXNUM\fP or \fB\*=inputfx=\fIFXNUM\fP
By default, multichannel play back defaults to fx8-fx13, this option
can be used to change the base fx offset (see multichannel.txt
distributed with the driver for more info).
.TP
\fB\-M\fIDEVICE\fP or \fB\*=mixer=\fIDEVICE\fP
Use a device other than /dev/mixer/
.TP
\fB\-p\fP or \fB\*=print\fP
prints mixer settings
.TP
\fB\-r\fP or \fB\*=reset=\fIBUS\fP
resets/prints the FX recording bus/es
.TP
\fB\-s\fP or \fB\*=source=\fISOURCE\fP
sets/prints the recording source
.TP
\fB\-?\fP or \fB\*=help\fP
Give this help list
.TP
\fB\*=usage
Give a short usage message
.TP
\fB\-V\fP or --version
	Print program version

.SH SHARED OUTPUT
.PP
One of the minijacks found on 5.1 series of SBLive cards as well as
Audigy Cards is a dual-mode analog/digital jack. It defaults to analog
mode, to switch it into digital mode, execute the following command:
.PP
.B emu-config -d
.PP
To switch it back to analog mode, use -a instead.
.PP
Note, OEM SBLive cards present in Gateway brand computers also have a
shared analog/digital jack but use a different and still unknown
method for switching the card.
.SH RECORDING MODES
.PP
There are three different recording modes on SBLive cards: ADC, MIC
and FX. ADC is the normal setting.
.TP
.B ADC mode:

This is the default setting. This mode can allow you to record at
different sample rates, with 1 or 2 channels. When selected, this mode will
cause the emu10k1 to copy the contents of what ever was sent to "ADC L/R"
through a sample rate converter before sending it to your computer and
eventually to the recording program.


By setting up a route to "ADC L/R" using emu-dspmgr, you can record
anything. (do not equate this with the analog AC97 mixer, it just so happens that we
have it set to that by default)


So if you wanted to record the cd digital input, you'd do:

(first remove analog sources)

.B emu-dspmgr -r"Analog L:ADC Rec L"

.B emu-dspmgr -r"Analog R:ADC Rec R"

(then add the cd spdif routes)

.B emu-dspmgr -a"CD-Spdif L:ADC Rec L"

.B emu-dspmgr -a"CD-Spdif R:ADC Rec R"


This mode probably does sample rate convertions even if you record at
48kHz (I should point out that internally, the emu10k1 does all it's work
at 48kHz), so you can probably never get a perfect copy using this method.


.TP
.B Mic mode:

This mode is intended as a "monitor" mode. Whatever is sent to "mic"
using dsp code, is copied by the emu10k1 through a sample rate convert to
8kHz (8 bit?) and off to your recording program.

To select mic recording:

.B emu-config -s"Mic"

This mode is kinda useless since the ADC record mode can also do mono
8bit, 8000Hz.

.TP
.B FX mode:


This mode allows for 16bit multi-track non-sample rate converted recording


Using FX recording, you can record any of the outputs listed in emu-dspmgr
(though the driver currently limits you to the first 16 for some
reason, need to verify and fix). Recording more than two channels is
possible, but the driver's OSS interface still incorrectly reports 2
channels of audio (the audio channels will be interleaved into 2 channels).


To select FX mode:

.B emu-config -s"FX"

You can see which outputs (or "buses") are selected:

.B emu-config -b
gives:

"FX bus[es] selected for recording: 0x0 0x1"

This means that the first two outputs,  0x0 (Front L) and 0x1 (Front R)
are selected for FX recording. If you were to record now, you'd record
whatever was being sent to those outputs by the dsp code.

To select different outputs (or "buses"), for example the digital outputs:

.B emu-config -b2
.B emu-config -b3

now 
.B emu-config -b
gives us :

"FX bus[es] selected for recording: 0x0 0x1 0x2 0x3"

so we need to remove 0x0 and 0x1 (or else we'll end up recording 4
channels of audio).

.B emu-config -r0
.B emu-config -r1

now:
"FX bus[es] selected for recording: 0x2 0x3"

You can now record what ever is being sent to the digital out.

A few things to note about FX mode:

- The drivers still reports (according to the oss query utility) that it
can record at different sampling rates. This is a bug, once in FX mode you
should only record at 48K, 16bit.

- As not all cards use all outputs, you can use the unused outputs on your
card as dummy outputs, and record from them too. This allows you to record
a different mix then what you hear. 


.SH REPORTING BUGS
Report bugs to <emu10k1-tools-devel@opensource.creative.com>

.SH COPYING
\(co 2001 Daniel Bertrand
.br
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translations approved by the copyright holders instead of in
the original English.
.SH SEE ALSO
emu_dspmgr(1), as10k1(1)
.SH AUTHORS
Daniel Bertrand <d.bertrand@ieee.org>







src/mixer/emu-config.c
/*********************************************************************
 *     cmd_line.c - a command line interface for the mixer library
 * 
 *      Copyright (C) 2000 Rui Sousa 
 ********************************************************************* 
 *     This program is free software; you can redistribute it and/or 
 *     modify it under the terms of the GNU General Public License as 
 *     published by the Free Software Foundation; either version 2 of 
 *     the License, or (at your option) any later version. 
 * 
 *     This program is distributed in the hope that it will be useful, 
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of 
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 *     GNU General Public License for more details. 
 * 
 *     You should have received a copy of the GNU General Public 
 *     License along with this program; if not, write to the Free 
 *     Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, 
 *     USA. 
 ********************************************************************* 
*/

#include <argp.h>
#include <stdlib.h>
#include <stdio.h>
#include "include/mix.h"
#include <string.h>
const char *argp_program_version = "mix 0.1";
const char *argp_program_bug_address = "<emu10k1-tools-devel@opensource.creative.com>";
const char doc[] = "mix - a program to control special emu10k1 mixer settings";

const struct argp_option options[] = {
	{"analog",'a', 0, 0,
	 "Set output in analog mode (5.1 cards only)"},
	{"bus", 'b', "BUS", OPTION_ARG_OPTIONAL,
	 "sets/prints the FX recording bus/es"},
	{"boost",'B',"on/off",0,
	 "Turn on or off a 12dB boost on analog front output (headphone mode)"},
	{"digital",'d', 0, 0,
	 "Set output in digital mode (5.1 cards only)"},
	{"spdifrate", 'f', "FREQ", 0,
	 "sets the output frequency in kHz of the S/PDIF port (use 44, 48 or 96)"},
	{"ir",'i', 0, 0,
	 "turn on IR remote control and midi port on the Live Drive"},
	{"info",'I', 0,0,"print card indo"},
	{"inputfx", 'm', "FXNUM", 0, 
	 "sets multichannel input fx"},
	{"mixer", 'M', "DEVICE", 0,
	 "use an alternate mixer device to /dev/mixer"},	
	{"print", 'p', 0, 0, "prints mixer settings"},
	{"reset", 'r', "BUS", OPTION_ARG_OPTIONAL,
	 "resets/prints the FX recording bus/es"},
	{"source", 's', "SOURCE", OPTION_ARG_OPTIONAL,
	 "sets/prints the recording source"},
	{"passthrough", 't', "ACTIVE", 0,
	 "sets digital pass-through mode and selects which S/PDIF ports to use"},
	{0}	
};

static error_t parser(int key, char *arg, struct argp_state *state)
{
	struct mix_settings *set = state->input;

	switch (key) {

	case 'a':
		set_51(set,0);
		break;
	case 'b':
		if (set->recsrc != mix_rec_source_number("FX"))
			argp_failure(state, EXIT_FAILURE, 0, "First select FX as the recording source");


		{
			int bus;
			if (arg) {
				sscanf(arg, "%d\n", &bus);
				if (mix_set_rec_fx_bus(set, bus) < 0)
					argp_failure(state, EXIT_FAILURE, 0, "Invalid recording FX bus");
			} else {
				printf("FX bus[es] selected for recording:");
				for (bus = 0; bus < 16; bus++)
					if (mix_check_rec_fx_bus(set, bus))
						printf(" 0x%1x", bus);
				printf("\n");
			}
		}
		break;
	case 'B':
		if(strncasecmp("on",arg, 2)==0)
			mix_ac97_boost(set,1);
		else
			mix_ac97_boost(set,0);
		break;
	case 'd':
		set_51(set,1);
		break;
	case 'f':
		mix_set_spdif_freq(set,atoi(arg));
		break;
	case 'i':
		enable_ir(set);
		break;
	case 'm':
		set_mch_fx(set, atoi(arg));
		break;
	case 'M':
		if(set->dev_name[0]!=0)
			break;
		strncpy(set->dev_name, arg, 64);
		set->dev_name[63]=0;
		break;
	case 't':
		set_passthrough(set, atoi(arg));
		break;
	case 'p':
		mix_print_settings(set);
		break;

	case 'r':
		if (set->recsrc != mix_rec_source_number("FX"))
			argp_failure(state, EXIT_FAILURE, 0, "First select FX as the recording source");


		{
			int bus;
			if (arg) {
				sscanf(arg, "%d\n", &bus);
				if (mix_reset_rec_fx_bus(set, bus) < 0)
					argp_failure(state, EXIT_FAILURE, 0, "Invalid recording FX bus");
			} else {
				printf("FX bus[es] not selected for recording:");
				for (bus = 0; bus < 0x10; bus++)
					if (!mix_check_rec_fx_bus(set, bus))
						printf(" 0x%1x", bus);
				printf("\n");
			}
		}
		break;

	case 's':
		if (arg) {
			if (mix_set_rec_source(set, arg) < 0)
				argp_failure(state, EXIT_FAILURE, 0, "Invalid recording source");
		} else {
			printf("Recording Source: %s\n", mix_get_rec_source(set));
		}
		break;
	case 'I':
		mix_print_info(set);
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

static struct argp argp = { options, parser, 0, doc, };

int main(int argc, char **argv)
{
	struct mix_settings set;
	int i;
	int ret;
	//We handle any -M option first, regardless of were it is
	set.dev_name[0]=0;
	for(i=0;i<argc;i++){
		if(strcmp(argv[i],"-M")==0 ||strcmp(argv[i],"--mixer")==0  ){
			if((argc>(i+1)) && (argv[i+1][0]!='-')){
				strncpy(set.dev_name, argv[i+1], 64);
				set.dev_name[63]=0;
			}
			goto match;
		}
	}

match:
	ret=mix_init(&set);
	if(ret)
		return ret;

	ret=argp_parse(&argp, argc, argv, 0, 0, &set);
	if(ret)
		return ret;
	
	ret=mix_commit(&set);
	
	return ret;
}
